pennylane/templates/subroutines/grover.py

```diff
@@ -0,0 +1,129 @@
+"""
+import pennylane as qml
+from pennylane.operation import AnyWires, Operation
+from pennylane.ops import Hadamard, PauliZ, MultiControlledX
+
+
+class GroverOperator(Operation):
+    r"""Performs the Grover Diffusion Operator.
+
+    .. math::
+
+        G = 2 |s \rangle \langle s | - I
+        = H^{\bigotimes n} \left( 2 |0\rangle \langle 0| - I \right) H^{\bigotimes n}
+
+    where :math:`n` is the number of wires, and :math:`|s\rangle` is the uniform superposition:
+
+    .. math::
+
+        |s\rangle = H^{\bigotimes n} |0\rangle =  \frac{1}{\sqrt{2^n}} \sum_{i=0}^{2^n} | i \rangle
+
+    For this template, the operator is implemented with a layer of Hadamards, an
+    effective multi-controlled Z gate, and another layer of Hadamards.
+
+    .. figure:: ../../_static/templates/subroutines/grover.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    This decomposition for the circuit is equivalent to:
+
+    .. figure:: ../../_static/templates/subroutines/grover2.svg
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    Args:
+        wires (Union[Wires, Sequence[int], or int]): the wires to apply to
+        work_wires (Union[Wires, Sequence[int], or int]): optional auxiliary wires to assist
+            in the decomposition of :class:`~.MultiControlledX`.
+
+    **Example**
+
+    The Grover Diffusion Operator amplifies the magnitude of the basis state with
+    a negative phase.  For example, if the solution to the search problem is the :math:`|111\rangle`
+    state, we require an oracle that flips its phase; this could be implemented using a `CCZ` gate:
+
+    .. code-block:: python
+        n_wires = 3
+        wires = list(range(n_wires))
+
+        def oracle():
+            qml.Hadamard(wires[-1])
+            qml.Toffoli(wires=wires)
+            qml.Hadamard(wires[-1])
+
+    We can then implement the entire Grover Search Algorithm for ``n`` iterations by alternating calls to the oracle and the diffusion operator:
+
+    .. code-block:: python
+
+        dev = qml.device('default.qubit', wires=wires)
+
+        @qml.qnode(dev)
+        def GroverSearch(num_iterations=1):
+            for wire in wires:
+                qml.Hadamard(wire)
+
+            for _ in range(num_iterations):
+                oracle()
+                qml.templates.GroverOperator(wires=wires)
+            return qml.probs(wires)
+
+    >>> GroverSearch(num_iterations=1)
+    tensor([0.03125, 0.03125, 0.03125, 0.03125, 0.03125, 0.03125, 0.03125,
+            0.78125], requires_grad=True)
+    >>> GroverSearch(num_iterations=2)
+    tensor([0.0078125, 0.0078125, 0.0078125, 0.0078125, 0.0078125, 0.0078125,
+        0.0078125, 0.9453125], requires_grad=True)
+
+    We can see that the marked :math:`|111\rangle` state has the greatest probability amplitude.
+
+    Optimally, the oracle-operator pairing should be repeated :math:`\frac{\pi}{4}\sqrt{2^{n}}$ times.
```

Comment:
The equation does not render well in the documentation.

Reply:
start in markdown, end in latex... oops